src/MSBuildLocator/DotNetSdkLocationHelper.cs

```diff
@@ -131,35 +131,51 @@ private static void ModifyUnmanagedDllResolver(Action<AssemblyLoadContext> resol
 
         private static IntPtr HostFxrResolver(Assembly assembly, string libraryName)
         {
+            // Library name for libhostfxr
             string hostFxrLibName = "libhostfxr";
+            // Library extension for the current platform
             string libExtension = RuntimeInformation.IsOSPlatform(OSPlatform.OSX) ? "dylib" : "so";
 
+            // If the requested library name is not libhostfxr, return IntPtr.Zero
             if (!hostFxrLibName.Equals(libraryName))
             {
                 return IntPtr.Zero;
             }
 
-            string hostFxrRoot = Path.Combine(s_dotnetPath.Value, "host", "fxr");
-            if (Directory.Exists(hostFxrRoot))
+            // Get the dotnet path candidates
+            foreach (string dotnetPath in s_dotnetPathCandidates.Value)
             {
-                // Load hostfxr from the highest version, because it should be backward-compatible
-                SemanticVersion? hostFxrAssemblyDirectory = Directory.GetDirectories(hostFxrRoot)
-                            .Max(str => SemanticVersionParser.TryParse(str, out SemanticVersion? version) ? version : null);
+                string hostFxrRoot = Path.Combine(dotnetPath, "host", "fxr");
 
-                if (hostFxrAssemblyDirectory != null && !string.IsNullOrEmpty(hostFxrAssemblyDirectory.OriginalValue))
+                // Check if the host/fxr directory exists
+                if (Directory.Exists(hostFxrRoot))
                 {
-                    string hostFxrAssembly = Path.Combine(hostFxrAssemblyDirectory.OriginalValue, Path.ChangeExtension(hostFxrLibName, libExtension));
-
-                    if (File.Exists(hostFxrAssembly))
+                    // Get a list of hostfxr assembly directories (e.g., 6.0.3, 7.0.1-preview.2.4)
+                    IList<SemanticVersion> hostFxrAssemblyDirs = Directory.GetDirectories(hostFxrRoot)
+                        .Select(path => SemanticVersionParser.TryParse(Path.GetFileName(path), out SemanticVersion? version) ? version : null)
+                        .Select(v => v!)
```

Comment:
The previous `Select` may return null. Shouldn't this be a `Where` filter?

Reply:
It's strange, but` .Where(v => v != null) `instead of ` .Select(v => v!)` generates [CS8619](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-messages/nullable-warnings?f1url=%3FappId%3Droslyn%26k%3Dk(CS8619)#mismatch-in-nullability-declaration:~:text=match%20implemented%20member.-,CS8619,-%2D%20Nullability%20of)

Reply:
It is tracked as a Roslyn issue: https://github.com/dotnet/roslyn/issues/37468